device/iim42652/calibration.go

```diff
@@ -0,0 +1,342 @@
+	err = i.WriteRegister(RegisterAccelGyroConfig, gyroAccelConfig)
+	if err != nil {
+		return err
+	}
+
+	time.Sleep(60 * time.Millisecond)
+	return nil
+}
+
+func (i *IIM42652) averageGyroSensorOutput() (average [3]int32, err error) {
+	numSamples := int32(0)
+	samplesDiscarded := int32(0)
+	timeout := 300 // us ?
+	sum := [3]int32{0, 0, 0}
+	for {
+		gyroscopeData, err := i.GetGyroscopeData()
+		if err != nil {
+			return average, err
+		}
+
+		if gyroscopeData.RawX != -32768 && gyroscopeData.RawY != -32768 && gyroscopeData.RawZ != -32768 {
+			sum[0] += int32(gyroscopeData.RawX)
+			sum[1] += int32(gyroscopeData.RawY)
+			sum[2] += int32(gyroscopeData.RawZ)
+		} else {
+			samplesDiscarded++
+		}
+
+		numSamples++
+		time.Sleep(time.Millisecond)
+		timeout--
+		if (numSamples == 200) || (timeout == 0) {
+			break
+		}
+	}
+
+	// compute average value
+	numSamples -= samplesDiscarded
+	average[0] = (sum[0] / numSamples)
+	average[1] = (sum[1] / numSamples)
+	average[2] = (sum[2] / numSamples)
+	return average, nil
+}
+
+// Negate the bias value and onvert the reading to the 64dps equivalent.
+func convertGyroBiasToRegisterFormat(gyroBias int32) int16 {
+	return int16(-(gyroBias * gyroOffuserConfiguredDps / gyroOffuserMaxDps) >> 4)
+}
+
+func (i *IIM42652) writeGyroBiasToUserRegister(bias [3]int32) error {
+	// The 3 bias values are stored as 12 bits each.
+	// They need to be stored interleaved across 5 byte registers
+	// Look at the IIM42652 datasheet for more info.
+	data := [5]byte{0, 0, 0, 0, 0}
+
+	// The accelerometer bias data shares a register with the gyrosocope.
+	// Copy the overlapping data to data[4] so that we don't lose it.
+	accelData, err := i.ReadRegister(RegisterOffsetUser4)
+	if err != nil {
+		return err
+	}
+	data[4] = (accelData & bitAccelXOffuserMaskHi)
+
+	cur_bias := convertGyroBiasToRegisterFormat(bias[0])
+	data[0] = storeLowBits(cur_bias, bitGyroXOffuserPosLo)
+	data[1] = storeHighBits(cur_bias, bitGyroXOffuserPosHi)
+
+	cur_bias = convertGyroBiasToRegisterFormat(bias[1])
+	data[1] |= storeHighBits(cur_bias, bitGyroYOffuserPosHi)
+	data[2] = storeLowBits(cur_bias, bitGyroYOffuserPosLo)
+
+	cur_bias = convertGyroBiasToRegisterFormat(bias[2])
+	data[3] = storeLowBits(cur_bias, bitGyroZOffuserPosLo)
+	data[4] |= storeHighBits(cur_bias, bitGyroZOffuserPosHi)
+
+	userRegister := *RegisterOffsetUser0
+	for idx := 0; idx < len(data); idx++ {
+		err := i.WriteRegister(&userRegister, data[idx])
+		if err != nil {
+			return err
+		}
+		userRegister.Address += 1
+	}
+
+	return nil
+}
+
+// Calibrates the gyro by taking an average and
+// storing the offset
+func (i *IIM42652) CalibrateGyro() (bias [3]int32, err error) {
+	err = i.initializeGyroForCalibration()
+	if err != nil {
+		return bias, err
+	}
+	time.Sleep(60 * time.Millisecond)
+
+	bias, err = i.averageGyroSensorOutput()
+	if err != nil {
+		return bias, err
+	}
+
+	err = i.writeGyroBiasToUserRegister(bias)
+	if err != nil {
+		return bias, err
+	}
+
```

Comment:
@punov Yes I think we can do this

One way might be to take another average after calibrating and verifying that those averages are close to 0,0,0.  That way we at least know that the offsets are working.

Reply:
Done, added basic calibration verification. 

When calibrating, we only give a warning to the user that the values don't seem right.  

With the `--verify-calibration` flag, the program will verify the current calibration (it will not recalibrate the device).  If verification fails, it will exit with code 1.